url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/SimpleTriggerImpl.java

```diff
@@ -0,0 +1,605 @@
+      return false;
+    }
+
+    if (misfireInstruction > MISFIRE_INSTRUCTION_RESCHEDULE_NEXT_WITH_EXISTING_COUNT) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * <p>
+   * Updates the <code>SimpleTrigger</code>'s state based on the MISFIRE_INSTRUCTION_XXX that was selected when the <code>SimpleTrigger</code> was
+   * created.
+   * </p>
+   * <p>
+   * If the misfire instruction is set to MISFIRE_INSTRUCTION_SMART_POLICY, then the following scheme will be used: <br>
+   * <ul>
+   * <li>If the Repeat Count is <code>0</code>, then the instruction will be interpreted as <code>MISFIRE_INSTRUCTION_FIRE_NOW</code>.</li>
+   * <li>If the Repeat Count is <code>REPEAT_INDEFINITELY</code>, then the instruction will be interpreted as
+   * <code>MISFIRE_INSTRUCTION_RESCHEDULE_NEXT_WITH_REMAINING_COUNT</code>. <b>WARNING:</b> using
+   * MISFIRE_INSTRUCTION_RESCHEDULE_NEXT_WITH_REMAINING_COUNT with a trigger that has a non-null end-time may cause the trigger to never fire again if
+   * the end-time arrived during the misfire time span.</li>
+   * <li>If the Repeat Count is <code>&gt; 0</code>, then the instruction will be interpreted as
+   * <code>MISFIRE_INSTRUCTION_RESCHEDULE_NOW_WITH_EXISTING_REPEAT_COUNT</code>.</li>
+   * </ul>
+   * </p>
+   * <p>
+   * If the misfire instruction is set to <code>Trigger.MISFIRE_INSTRUCTION_SKIP_TO_NEXT_FIRE_AFTER_CURRENT_DATE</code> then the behavior will be
+   * identical to
+   */
+  @Override
+  public void updateAfterMisfire(Calendar cal) {
+
+    int instr = getMisfireInstruction();
+
+    if (instr == Trigger.MISFIRE_INSTRUCTION_IGNORE_MISFIRE_POLICY) {
+      return;
+    }
+
+    if (instr == Trigger.MISFIRE_INSTRUCTION_SMART_POLICY) {
+      if (getRepeatCount() == 0) {
+        instr = MISFIRE_INSTRUCTION_FIRE_NOW;
+      } else if (getRepeatCount() == REPEAT_INDEFINITELY) {
+        instr = MISFIRE_INSTRUCTION_RESCHEDULE_NEXT_WITH_REMAINING_COUNT;
+      } else {
+        // if (getRepeatCount() > 0)
+        instr = MISFIRE_INSTRUCTION_RESCHEDULE_NOW_WITH_EXISTING_REPEAT_COUNT;
+      }
+    } else if (instr == MISFIRE_INSTRUCTION_FIRE_NOW && getRepeatCount() != 0) {
+      instr = MISFIRE_INSTRUCTION_RESCHEDULE_NOW_WITH_REMAINING_REPEAT_COUNT;
+    }
+
+    if (instr == MISFIRE_INSTRUCTION_FIRE_NOW) {
+      setNextFireTime(new Date());
+    } else if (instr == MISFIRE_INSTRUCTION_RESCHEDULE_NEXT_WITH_EXISTING_COUNT) {
+      Date newFireTime = getFireTimeAfter(new Date());
+      while (newFireTime != null && cal != null && !cal.isTimeIncluded(newFireTime.getTime())) {
+        newFireTime = getFireTimeAfter(newFireTime);
+
+        if (newFireTime == null) {
+          break;
+        }
+
+        // avoid infinite loop
+        java.util.Calendar c = java.util.Calendar.getInstance();
+        c.setTime(newFireTime);
+        if (c.get(java.util.Calendar.YEAR) > YEAR_TO_GIVEUP_SCHEDULING_AT) {
+          newFireTime = null;
+        }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 267:280, 284:297, 356:370.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.